deps/ipc-util/ipc-util/pipe.h

```diff
@@ -33,16 +33,39 @@ typedef struct ipc_pipe_server ipc_pipe_server_t;
 typedef struct ipc_pipe_client ipc_pipe_client_t;
 
 typedef void (*ipc_pipe_read_t)(void *param, uint8_t *data, size_t size);
+typedef void (*ipc_pipe_connect_t)(void *param);
+typedef bool (*ipc_pipe_disconnect_t)(void *param);
 
 bool ipc_pipe_server_start(ipc_pipe_server_t *pipe, const char *name,
 			   ipc_pipe_read_t read_callback, void *param);
+void ipc_pipe_server_set_connect_callback(ipc_pipe_server_t *pipe,
+					  ipc_pipe_connect_t connect_callback,
+					  void *param);
+void ipc_pipe_server_set_disconnect_callback(
+	ipc_pipe_server_t *pipe, ipc_pipe_disconnect_t disconnect_callback,
+	void *param);
+void ipc_pipe_server_disconnect(ipc_pipe_server_t *pipe);
 void ipc_pipe_server_free(ipc_pipe_server_t *pipe);
+bool ipc_pipe_server_write(ipc_pipe_server_t *pipe, const void *data,
+			   size_t size);
```

Comment:
The only time the obs-ipc plugin writes on a pipe it also reads from is the ResponseServer, where it waits on a connect message, and responds with information about an incoming and outgoing pipe before disconnecting, or just drops the connection if a connect message isn't sent.

Since the client sends a message and waits on a response, and the server is waiting on a message to then send a response, with both disconnecting immediately after, there is no chance of a lock up as far as I am aware. All other information and signals sent are done through two single direction pipes.

Reply:
Indeed, so basically `TransactNamedPipe`; you probably want to add a comment to `ipc_pipe_server_write` to detail that it won't work properly outside of this very specific usage pattern